jsonComposer.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2024-Present Couchbase, Inc. All rights reserved.
+
+package gojsonsm
+
+import (
+	"fmt"
+)
+
+var (
+	ErrNilComposer        error = fmt.Errorf("nil composer")
+	ErrNilData            error = fmt.Errorf("nil data")
+	ErrInvalidJSON        error = fmt.Errorf("invalid JSON object")
+	ErrUnexpectedEOF      error = fmt.Errorf("unexpected EOF")
+	ErrInsufficientMemory error = fmt.Errorf("insufficient memory allocated for dst, cannot proceed")
+	ErrUnrecognisedToken  error = fmt.Errorf("unrecognised token in the JSON object")
+)
+
+type jsonObjComposer struct {
+	// should have enough length
+	body []byte
+	// cursor
+	pos int
+	// if the newly composed json has atleast one field
+	atLeastOneFieldLeft bool
+}
+
+// writes a given slice of data to the composer
+// returns any errors in the process
+func (composer *jsonObjComposer) Write(data []byte, tknType tokenType) error {
+	if composer == nil {
+		return ErrNilComposer
+	}
+	if data == nil {
+		return ErrNilData
+	}
+	// if we are about to write objectEnd token i.e. "}", we should ensure that the previous position was not a ","
+	// if it is ",", step back by one position and write a "}"
+	if tknType == tknObjectEnd && composer.pos-1 >= 0 && composer.body[composer.pos-1] == ',' {
+		composer.pos--
+	}
+	n := copy(composer.body[composer.pos:composer.pos+len(data)], data)
+	if n != len(data) {
+		return ErrInsufficientMemory
+	}
+	composer.pos += len(data)
+	return nil
+}
+
+// finalizes the composer data and sets atLeastOneFieldLeft
+// returns (length of composer data, if atleast one field is left in the newly composed JSON object)
+func (composer *jsonObjComposer) Commit() (int, bool, error) {
+	if composer == nil {
+		return 0, false, ErrNilComposer
+	}
+	// if the body is anything other than {}, we have atleast one field inside it
+	composer.atLeastOneFieldLeft = composer.pos > 2
+
+	return composer.pos, composer.atLeastOneFieldLeft, nil
+}
+
+func handleError(err error) (int, int, bool, error) {
+	return 0, 0, false, err
+}
+
+// Given a byte encoded json object - "src", a list of keys of items to remove from src - "remove",
+// the function removes the items from "src" and places them in "removed" and the remaining json object is stored in "dst".
+// It returns (final length of dst, number of items removed, if there are any items left in dst at the end, error).
+// Caller has the ability to pass in pre-allocated byte slices for dst and pre-allocated map for removed. If nil is passed, only then memory is allocated.
+func MatchAndRemoveItemsFromJsonObject(src []byte, remove []string, dst []byte, removed map[string][]byte) (int, int, bool, error) {
+	if len(src) < 2 || src[0] != '{' || src[len(src)-1] != '}' {
```

Comment:
Related to the above, it also might be worth considering using the same sort of 'precalculation' step that jsonsm  uses, where you could have `remove` be pre-parsed into a tree that allows the actual execution step to quickly determine whether a particular key can be entirely skipped, or needs to perform nested checking.  This would pull any 'parsing' side of things out of jsonsm itself, and jsonsm would continue to strictly operate at the byte-level.

Reply:
`remove` is actually a list of keys which can be directly "matched" 1:1 with the JSON keys. So I don't see a necessity of any more pre-parsing/pre-calculation.

> actual execution step to quickly determine whether a particular key can be entirely skipped, or needs to perform nested checking
This is what the function is currently doing. 
1. get the next token
2. if the token is a JSON key (nested or otherwise), then try to exactly match the token with a string in `remove`
3. if it doesn't match, do to step (1)
4. if it matches, then don't write it and its corresponding value to the new composed object